napari/utils/lock.py

```diff
@@ -0,0 +1,215 @@
+                    "Value should be numeric for lock modes: LARGER_THAN and SMALLER_THAN",
+                    deferred=True,
+                ),
+            )
+
+        if value_lock_mode == LockMode.IN_LIST and type(value) not in (
+            list,
+            tuple,
+        ):
+            raise ValueNotCompatibleWithLockMode(
+                value=value,
+                message=trans._(
+                    "Value should be list or tuple for lock mode: IN_LIST",
+                    deferred=True,
+                ),
+            )
+
+        if value_lock_mode == LockMode.IN_RANGE and (
+            type(value) not in (list, tuple) or len(value) != 2
+        ):
+            raise ValueNotCompatibleWithLockMode(
+                value=value,
+                message=trans._(
+                    "Value should be list or tuple of 2 elements for lock mode: IN_RANGE",
+                    deferred=True,
+                ),
+            )
+
+        return value
+
+    def lock(self) -> None:
+        """locks the lock"""
+        self.locked = True
+
+    def unlock(self) -> None:
+        """locks the lock"""
+        self.locked = False
+
+    def is_valid_value(self, value: Any) -> bool:
+        """Checks if value is permitted in the lock"""
+        if self.value_lock_mode == LockMode.EXACT:
+            return False
+        elif self.value_lock_mode == LockMode.IN_LIST:
+            return value in self.value
+        elif self.value_lock_mode == LockMode.IN_RANGE:
+            return value >= self.value[0] and value <= self.value[1]
+        elif self.value_lock_mode == LockMode.SMALLER_THAN:
+            return value <= self.value
+        # elif self.value_lock_mode == LockMode.LARGER_THAN:
+        return value >= self.value
+
+    def is_locked(self, hard_lock: bool = True, value: Any = None):
+        """Check if there is a lock on a specific attribute"""
+        if not self.locked:
+            return False
+
+        if self.hard_lock is True or hard_lock is False:
+            if self.is_valid_value(value):
+                return False
+            else:
+                return True
+
+        return False
+
+
+class Locker:
+    """
+    A generic class to be used as a locking mechanism. It contains a collection of locks
+    for different attributes.
+    """
+
+    _lock_dictionary: Dict[str, Lock]
+
+    def __init__(self, lock_dictionary: Optional[Dict[str, Lock]] = None):
+        """Initializes the locker"""
+        if lock_dictionary is not None:
+            self._lock_dictionary = lock_dictionary
+        else:
+            self._lock_dictionary = {}
+
+    def add_lock(
+        self,
+        attribute: str,
+        value: Any = None,
+        hard_lock: bool = True,
+        locked: bool = True,
+        value_lock_mode: LockMode = LockMode.EXACT,
+        comments: Optional[str] = None,
+    ) -> None:
+        """Adds a lock to the locker"""
+        self._lock_dictionary[attribute] = Lock(
```

Comment:
Should there be any check that attribute is not already set and/or locked ?

Reply:
Added a check with error if it is already added.